### PR TITLE
Update tempo hack script for OpenShift

### PR DESCRIPTION
**Pull request by josunect**
Fixes: https://github.com/kiali/kiali/issues/6663 

Run: 

`istio/tempo/install-tempo-env.sh -c oc`

Kiali config: 

```
  tracing:
    enabled: true
    provider: "tempo"
    in_cluster_url: http://localhost:3200
    url: http://tempo-cr-query-frontend-tempo.apps-crc.testing
    use_grpc: false
```

image

### CODE REVIEW

File: hack/istio/tempo/install-tempo-env.sh [hunk 1]
```diff
@@ -80,6 +80,23 @@ done
...
+  else
+    echo "ERROR: You must install the cluster client ${CLIENT_EXE_NAME} in your PATH before you can continue."
+    exit 1
+  fi
+fi
+
+IS_OPENSHIFT="false"
+if [[ "${CLIENT_EXE}" = *"oc" ]]; then
+  IS_OPENSHIFT="true"
+fi
```
Comment (jmazzitelli) [inline hack/istio/tempo/install-tempo-env.sh:96]: I recently changed our other hack script to be more specific when determining if the server is OpenShift or not. 

We should consider doing this in our hack scripts now:

https://github.com/kiali/kiali/blob/master/hack/istio/install-istio-via-istioctl.sh#L281-L287

```suggestion
if ${CLIENT_EXE} api-versions | grep --quiet "route.openshift.io"; then
  IS_OPENSHIFT="true"
  echo "You are connecting to an OpenShift cluster"
else
  IS_OPENSHIFT="false"
  echo "You are connecting to a (non-OpenShift) Kubernetes cluster"
fi
```

File: hack/istio/tempo/install-tempo-env.sh [hunk 2]
```diff
@@ -9,7 +9,7 @@
 #
 ##############################################################################
 
-CLIENT_EXE="kubectl"
+CLIENT_EXE="oc"
```
Comment (jshaughn) [inline hack/istio/tempo/install-tempo-env.sh:12]:
```suggestion
CLIENT_EXE_NAME="oc"
```

I think this is better handled the same way as in the bookinfo install script.

File: hack/istio/tempo/install-tempo-env.sh [hunk 3]
```diff
@@ -160,6 +191,11 @@ EOF
     ${SCRIPT_DIR}/../install-bookinfo-demo.sh -c ${CLIENT_EXE} -tg
   fi
 
+  # If OpenShift, we need to do some additional things
+  if [ "${IS_OPENSHIFT}" == "true" ]; then
+    $CLIENT_EXE expose svc/tempo-cr-query-frontend -n ${TEMPO_NS}
```
Comment (jshaughn) [inline hack/istio/tempo/install-tempo-env.sh:196]:
```suggestion
    $CLIENT_EXE expose svc/tempo-cr-query-frontend -n ${TEMPO_NS}
    $CLIENT_EXE expose svc/grafana -n istio-system
```

File: hack/istio/tempo/install-tempo-env.sh [hunk 4]
```diff
@@ -160,6 +191,11 @@ EOF
...
+  fi
+
   echo -e "Installation finished. You can port forward the services with: \n"
   echo "./run-kiali.sh -pg 13000:3000 -pp 19090:9090 -pt 3200:3200 -app 8080 -es false -iu http://127.0.0.1:15014 -tr tempo-cr-query-frontend -ts tempo-cr-query-frontend -tn tempo"
```
Comment (jshaughn) [inline hack/istio/tempo/install-tempo-env.sh:200]:
```suggestion
  if [ "${IS_OPENSHIFT}" != "true" ]; then
    echo "./run-kiali.sh -pg 13000:3000 -pp 19090:9090 -pt 3200:3200 -app 8080 -es false -iu http://127.0.0.1:15014 -tr tempo-cr-query-frontend -ts tempo-cr-query-frontend -tn tempo"
  fi
```

**Review by jshaughn — COMMENTED** [review]
Testing on CRC (4.12) and some stuff is working. I have things defined like this in the Kiali CR:
```
spec:
  external_services:
    grafana:
      enabled: true
      in_cluster_url: http://grafana.istio-system:3000
      url: http://grafana-istio-system.apps-crc.testing
    tracing:
      enabled: true
      in_cluster_url: http://tempo-cr-query-frontend.tempo:3200
      provider: tempo
      url: http://tempo-cr-query-frontend-tempo.apps-crc.testing
      use_grpc: false
```

And the Tempo datasource is:
```
  is_default: on
  url: http://tempo-cr-query-frontend.tempo:3200  // same as in_cluster_url

```

I'm running with Kiali master, in cluster (not run-kiali).  I defined routes for tempo query frontend and grafana:
```
$ oc get routes -n istio-system
NAME                   HOST/PORT                                            PATH   SERVICES               PORT      TERMINATION          WILDCARD
grafana                grafana-istio-system.apps-crc.testing                       grafana                service                        None
$ oc get routes -n tempo
NAME                      HOST/PORT                                        PATH   SERVICES                  PORT   TERMINATION   WILDCARD
tempo-cr-query-frontend   tempo-cr-query-frontend-tempo.apps-crc.testing          tempo-cr-query-frontend   http                 None
```

A couple of issues:

1. The url for the main external link to Distributed Tracing sends me to the url defined for tempo which returns 404.  Is that expected?  Should it be pointing to grafana instead or possibly have some query params?
2. The trace scatterplot (Traces tab for a detail page) no longer fetches data when the mouse drags over it.  Instead it forces a click on the tooltip.  Was this an intentional change?  For Jaeger we would perform network requests to pull data (while ensuring we didn't launch concurrent requests).
3. After selecting a trace from the scatterplot, and selecting the `Spans` tab,  the "Load statistics` links don't seem to do anything (links found in some of the rows' `Statistics` column)
4. I think maybe the default for INSTALL_KIALI should be false (at least for now) because it will pull the version associated with the istio version (1.72 for 1.19).  I think we need to currently run a dev version.

I'm not sure what I should currently expect to work, and what may be a problem.  We will need to document things well for configuring for CRC, I think.

**Comment by josunect** [comment]
> Testing on CRC (4.12) and some stuff is working. I have things defined like this in the Kiali CR:
> 
> ```
> spec:
>   external_services:
>     grafana:
>       enabled: true
>       in_cluster_url: http://grafana.istio-system:3000
>       url: http://grafana-istio-system.apps-crc.testing
>     tracing:
>       enabled: true
>       in_cluster_url: http://tempo-cr-query-frontend.tempo:3200
>       provider: tempo
>       url: http://tempo-cr-query-frontend-tempo.apps-crc.testing
>       use_grpc: false
> ```
> 
> And the Tempo datasource is:
> 
> ```
>   is_default: on
>   url: http://tempo-cr-query-frontend.tempo:3200  // same as in_cluster_url
> ```
> 
> I'm running with Kiali master, in cluster (not run-kiali). I defined routes for tempo query frontend and grafana:
> 
> ```
> $ oc get routes -n istio-system
> NAME                   HOST/PORT                                            PATH   SERVICES               PORT      TERMINATION          WILDCARD
> grafana                grafana-istio-system.apps-crc.testing                       grafana                service                        None
> $ oc get routes -n tempo
> NAME                      HOST/PORT                                        PATH   SERVICES                  PORT   TERMINATION   WILDCARD
> tempo-cr-query-frontend   tempo-cr-query-frontend-tempo.apps-crc.testing          tempo-cr-query-frontend   http                 None
> ```
> 
> A couple of issues:
> 
>     1. The url for the main external link to Distributed Tracing sends me to the url defined for tempo which returns 404.  Is that expected?  Should it be pointing to grafana instead or possibly have some query params?

Yes, I was not sure what to put in there but I think it makes more sense. I will change that in a related task. 
> 
>     2. The trace scatterplot (Traces tab for a detail page) no longer fetches data when the mouse drags over it.  Instead it forces a click on the tooltip.  Was this an intentional change?  For Jaeger we would perform network requests to pull data (while ensuring we didn't launch concurrent requests).

I was looking at the Jaeger behavior to have the same, and it looks like it just makes an Trace details request on click. The difference is that for Tempo, we don't have all the Trace info in the search request. But, we can change the event to hover if it makes more sense. 
> 
>     3. After selecting a trace from the scatterplot, and selecting the `Spans` tab,  the "Load statistics`links don't seem to do anything (links found in some of the rows'`Statistics` column)
> 
I see a warning in the console, `Could not determine direction from Envoy span` I think it _might_ be related to some data missing, like https://github.com/kiali/kiali/issues/6674, I will investigate. 

>     4. I think maybe the default for INSTALL_KIALI should be false (at least for now) because it will pull the version associated with the istio version (1.72 for 1.19).  I think we need to currently run a dev version.
> 
Right, I have disabled it by default. 

> 
> I'm not sure what I should currently expect to work, and what may be a problem. We will need to document things well for configuring for CRC, I think.

+1

**Review by jshaughn — APPROVED** [review]
@josunect Thanks for the changes, approving the script, the overall work is still in progress and we can make more changes as we go, if needed.

> 1. The url for the main external link to Distributed Tracing sends me to the url defined for tempo which returns 404.  Is that expected?  Should it be pointing to grafana instead or possibly have some query params?

> Yes, I was not sure what to put in there but I think it makes more sense. I will change that in a related task. https://github.com/kiali/kiali/issues/6687 

Thumbs up!

> 2. The trace scatterplot (Traces tab for a detail page) no longer fetches data when the mouse drags over it.  Instead it forces a click on the tooltip.  Was this an intentional change?  For Jaeger we would perform network requests to pull data (while ensuring we didn't launch concurrent requests).

> I was looking at the Jaeger behavior to have the same, and it looks like it just makes an Trace details request on click. The difference is that for Tempo, we don't have all the Trace info in the search request. But, we can change the event to hover if it makes more sense.

If you look at the Jaeger behavior, by opening the browser debug and watching Network requests, you can see that as the mouse passes over points on the chart, it pulls information automatically, and so when a user rests the cursor over a point the TT should be populated.  I think we want to keep this behavior, if possible.  It is not as nice of an experience to have  to click.   We can discuss further in a separate issue.

> 3. After selecting a trace from the scatterplot, and selecting the `Spans` tab,  the "Load statistics`links don't seem to do anything (links found in some of the rows'`Statistics` column)

> I see a warning in the console, Could not determine direction from Envoy span I think it might be related to some data missing, like https://github.com/kiali/kiali/issues/6674, I will investigate.

Thumbs up!

> I'm not sure what I should currently expect to work, and what may be a problem. We will need to document things well for configuring for CRC, I think.

> https://github.com/kiali/kiali/issues/6689

Thumbs Up!

**Pull request merged**